### PR TITLE
Rewrite serializer for work with custom attributes

### PR DESCRIPTION
**Pull request by myxaluch**
Немного переписал сериалайзер, что бы была возможность работать с кастомными методами в самом сериалайзере.

@krim @jbmeerkat Посмотрите, как будет время

UPD. Проверил на бенчмарке вот отсюда (https://github.com/nesaulov/surrealist/blob/master/benchmarks/surrealist_vs_ams.rb, добавил его в этом же PR)
когда была проверка типов - это было чертовски медленно, медленнее, чем AMS. 
Убрал проверку типов - стало просто как пушка, привожу результаты:
Warming up --------------------------------------
      AMS : instance     2.603k i/100ms
Surrealist: instance through .surrealize
                         3.969k i/100ms
Light: instance through .to_json
                         9.510k i/100ms
Surrealist: instance through Surrealist::Serializer
                         3.265k i/100ms
ActiveModel::Serializers::JSON instance
                         3.059k i/100ms
         Blueprinter     5.711k i/100ms
Calculating -------------------------------------
      AMS : instance     26.848k (±11.2%) i/s -    132.753k in   5.018274s
Surrealist: instance through .surrealize
                         48.517k (± 6.1%) i/s -    242.109k in   5.011215s
Light: instance through .to_json
                        107.351k (± 6.8%) i/s -    542.070k in   5.075870s
Surrealist: instance through Surrealist::Serializer
                         32.275k (± 9.1%) i/s -    159.985k in   5.006392s
ActiveModel::Serializers::JSON instance
                         27.487k (±13.1%) i/s -    137.655k in   5.115311s
         Blueprinter     54.179k (±11.8%) i/s -    268.417k in   5.054407s

Comparison:
Light: instance through .to_json:   107351.4 i/s
         Blueprinter:    54179.0 i/s - 1.98x  slower
Surrealist: instance through .surrealize:    48516.6 i/s - 2.21x  slower
Surrealist: instance through Surrealist::Serializer:    32274.8 i/s - 3.33x  slower
ActiveModel::Serializers::JSON instance:    27486.6 i/s - 3.91x  slower
      AMS : instance:    26847.8 i/s - 4.00x  slower

Warming up --------------------------------------
    AMS : collection     1.000  i/100ms
Surrealist: collection through Surrealist.surrealize_collection()
                         1.000  i/100ms
Light: collection through .to_json
                         4.000  i/100ms
Surrealist: collection through Surrealist::Serializer
                         1.000  i/100ms
ActiveModel::Serializers::JSON collection
                         1.000  i/100ms
Blueprinter collection
                         3.000  i/100ms
Calculating -------------------------------------
    AMS : collection     12.009  (± 8.3%) i/s -     60.000  in   5.018682s
Surrealist: collection through Surrealist.surrealize_collection()
                         17.940  (± 5.6%) i/s -     90.000  in   5.049161s
Light: collection through .to_json
                         50.631  (± 7.9%) i/s -    252.000  in   5.021358s
Surrealist: collection through Surrealist::Serializer
                         13.289  (±15.1%) i/s -     66.000  in   5.077054s
ActiveModel::Serializers::JSON collection
                         12.963  (± 7.7%) i/s -     65.000  in   5.071031s
Blueprinter collection
                         38.583  (±15.6%) i/s -    189.000  in   5.056796s

Comparison:
Light: collection through .to_json:       50.6 i/s
Blueprinter collection:       38.6 i/s - 1.31x  slower
Surrealist: collection through Surrealist.surrealize_collection():       17.9 i/s - 2.82x  slower
Surrealist: collection through Surrealist::Serializer:       13.3 i/s - 3.81x  slower
ActiveModel::Serializers::JSON collection:       13.0 i/s - 3.91x  slower
    AMS : collection:       12.0 i/s - 4.22x  slower

------- Turning off AMS logger -------
Warming up --------------------------------------
AMS(without logging) : instance
                         3.183k i/100ms
Surrealist: instance through .surrealize
                         5.071k i/100ms
Light: instance through .to_json
                        11.022k i/100ms
Surrealist: instance through Surrealist::Serializer
                         3.333k i/100ms
ActiveModel::Serializers::JSON instance
                         2.939k i/100ms
         Blueprinter     6.655k i/100ms
Calculating -------------------------------------
AMS(without logging) : instance
                         32.627k (± 2.7%) i/s -    165.516k in   5.076772s
Surrealist: instance through .surrealize
                         52.812k (± 4.2%) i/s -    263.692k in   5.002762s
Light: instance through .to_json
                        120.677k (± 2.4%) i/s -    606.210k in   5.026312s
Surrealist: instance through Surrealist::Serializer
                         36.909k (± 3.3%) i/s -    186.648k in   5.062296s
ActiveModel::Serializers::JSON instance
                         34.207k (± 3.4%) i/s -    173.401k in   5.075142s
         Blueprinter     67.828k (± 4.8%) i/s -    339.405k in   5.016120s

Comparison:
Light: instance through .to_json:   120676.7 i/s
         Blueprinter:    67828.1 i/s - 1.78x  slower
Surrealist: instance through .surrealize:    52812.3 i/s - 2.29x  slower
Surrealist: instance through Surrealist::Serializer:    36909.2 i/s - 3.27x  slower
ActiveModel::Serializers::JSON instance:    34206.8 i/s - 3.53x  slower
AMS(without logging) : instance:    32627.5 i/s - 3.70x  slower

Warming up --------------------------------------
AMS(without logging) : collection
                         1.000  i/100ms
Surrealist: collection through Surrealist.surrealize_collection()
                         1.000  i/100ms
Light: collection through .to_json
                         5.000  i/100ms
Surrealist: collection through Surrealist::Serializer
                         1.000  i/100ms
ActiveModel::Serializers::JSON collection
                         1.000  i/100ms
Blueprinter collection
                         4.000  i/100ms
Calculating -------------------------------------
AMS(without logging) : collection
                         12.444  (± 0.0%) i/s -     63.000  in   5.066429s
Surrealist: collection through Surrealist.surrealize_collection()
                         16.929  (±17.7%) i/s -     82.000  in   5.031321s
Light: collection through .to_json
                         51.883  (± 7.7%) i/s -    260.000  in   5.056180s
Surrealist: collection through Surrealist::Serializer
                         15.427  (± 6.5%) i/s -     77.000  in   5.028085s
ActiveModel::Serializers::JSON collection
                         14.453  (± 6.9%) i/s -     73.000  in   5.063155s
Blueprinter collection
                         39.911  (± 2.5%) i/s -    200.000  in   5.018043s

Comparison:
Light: collection through .to_json:       51.9 i/s
Blueprinter collection:       39.9 i/s - 1.30x  slower
Surrealist: collection through Surrealist.surrealize_collection():       16.9 i/s - 3.06x  slower
Surrealist: collection through Surrealist::Serializer:       15.4 i/s - 3.36x  slower
ActiveModel::Serializers::JSON collection:       14.5 i/s - 3.59x  slower
AMS(without logging) : collection:       12.4 i/s - 4.17x  slower

Warming up --------------------------------------
AMS (associations): instance
                       487.000  i/100ms
Surrealist (associations): instance through .surrealize
                       789.000  i/100ms
Light (associations): instance through .to_json
                         3.012k i/100ms
Surrealist (associations): instance through Surrealist::Serializer
                       762.000  i/100ms
ActiveModel::Serializers::JSON (associations)
                       930.000  i/100ms
Blueprinter (associations)
                         2.115k i/100ms
Calculating -------------------------------------
AMS (associations): instance
                          4.354k (±17.8%) i/s -     21.428k in   5.130085s
Surrealist (associations): instance through .surrealize
                          6.513k (±20.0%) i/s -     31.560k in   5.082532s
Light (associations): instance through .to_json
                         30.044k (± 9.1%) i/s -    150.600k in   5.065154s
Surrealist (associations): instance through Surrealist::Serializer
                          7.951k (± 3.4%) i/s -     40.386k in   5.084985s
ActiveModel::Serializers::JSON (associations)
                          8.836k (±15.6%) i/s -     42.780k in   5.009103s
Blueprinter (associations)
                         18.598k (±18.5%) i/s -     90.945k in   5.094459s

Comparison:
Light (associations): instance through .to_json:    30044.1 i/s
Blueprinter (associations):    18597.9 i/s - 1.62x  slower
ActiveModel::Serializers::JSON (associations):     8835.7 i/s - 3.40x  slower
Surrealist (associations): instance through Surrealist::Serializer:     7951.4 i/s - 3.78x  slower
Surrealist (associations): instance through .surrealize:     6513.3 i/s - 4.61x  slower
AMS (associations): instance:     4353.7 i/s - 6.90x  slower

Warming up --------------------------------------
AMS (associations): collection
                         1.000  i/100ms
Surrealist (associations): collection through Surrealist.surrealize_collection()
                         1.000  i/100ms
Light (associations): collection through .to_json
                         1.000  i/100ms
Surrealist (associations): collection through Surrealist::Serializer
                         1.000  i/100ms
ActiveModel::Serializers::JSON (associations): collection
                         1.000  i/100ms
Blueprinter (associations): collection
                         1.000  i/100ms
Calculating -------------------------------------
AMS (associations): collection
                          2.231  (± 0.0%) i/s -     11.000  in   5.019610s
Surrealist (associations): collection through Surrealist.surrealize_collection()
                          3.859  (±25.9%) i/s -     19.000  in   5.071439s
Light (associations): collection through .to_json
                         15.910  (±18.9%) i/s -     78.000  in   5.062683s
Surrealist (associations): collection through Surrealist::Serializer
                          4.245  (± 0.0%) i/s -     21.000  in   5.025093s
ActiveModel::Serializers::JSON (associations): collection
                          6.045  (±16.5%) i/s -     30.000  in   5.167062s
Blueprinter (associations): collection
                         15.206  (±13.2%) i/s -     75.000  in   5.003640s

Comparison:
Light (associations): collection through .to_json:       15.9 i/s
Blueprinter (associations): collection:       15.2 i/s - same-ish: difference falls within error
ActiveModel::Serializers::JSON (associations): collection:        6.0 i/s - 2.63x  slower
Surrealist (associations): collection through Surrealist::Serializer:        4.2 i/s - 3.75x  slower
Surrealist (associations): collection through Surrealist.surrealize_collection():        3.9 i/s - 4.12x  slower
AMS (associations): collection:        2.2 i/s - 7.13x  slower

------- Enabling Oj.optimize_rails() & Blueprinter config.generator = Oj -------
Warming up --------------------------------------
AMS(without logging) (with Oj): instance
                         5.508k i/100ms
Surrealist: instance through .surrealize
                         3.986k i/100ms
Light: instance through .to_json
                        10.580k i/100ms
Surrealist: instance through Surrealist::Serializer
                         2.887k i/100ms
ActiveModel::Serializers::JSON(with Oj) instance
                         5.070k i/100ms
Blueprinter(with Oj)     8.549k i/100ms
Calculating -------------------------------------
AMS(without logging) (with Oj): instance
                         63.732k (± 7.7%) i/s -    319.464k in   5.040069s
Surrealist: instance through .surrealize
                         50.630k (± 7.5%) i/s -    255.104k in   5.069889s
Light: instance through .to_json
                        114.978k (± 5.5%) i/s -    581.900k in   5.076195s
Surrealist: instance through Surrealist::Serializer
                         35.744k (± 6.0%) i/s -    178.994k in   5.026879s
ActiveModel::Serializers::JSON(with Oj) instance
                         58.034k (± 8.8%) i/s -    288.990k in   5.022256s
Blueprinter(with Oj)     84.562k (± 9.6%) i/s -    418.901k in   5.006254s

Comparison:
Light: instance through .to_json:   114978.4 i/s
Blueprinter(with Oj):    84562.2 i/s - 1.36x  slower
AMS(without logging) (with Oj): instance:    63732.1 i/s - 1.80x  slower
ActiveModel::Serializers::JSON(with Oj) instance:    58033.9 i/s - 1.98x  slower
Surrealist: instance through .surrealize:    50630.0 i/s - 2.27x  slower
Surrealist: instance through Surrealist::Serializer:    35744.0 i/s - 3.22x  slower

Warming up --------------------------------------
AMS(without logging) (with Oj): collection
                         1.000  i/100ms
Surrealist: collection through Surrealist.surrealize_collection()
                         1.000  i/100ms
Light: collection through .to_json
                         4.000  i/100ms
Surrealist: collection through Surrealist::Serializer
                         1.000  i/100ms
ActiveModel::Serializers::JSON(with Oj) collection
                         1.000  i/100ms
Blueprinter collection(with Oj)
                         3.000  i/100ms
Calculating -------------------------------------
AMS(without logging) (with Oj): collection
                         18.430  (±16.3%) i/s -     90.000  in   5.008760s
Surrealist: collection through Surrealist.surrealize_collection()
                         16.551  (±18.1%) i/s -     79.000  in   5.009888s
Light: collection through .to_json
                         45.879  (±17.4%) i/s -    224.000  in   5.059959s
Surrealist: collection through Surrealist::Serializer
                         13.547  (±14.8%) i/s -     67.000  in   5.068036s
ActiveModel::Serializers::JSON(with Oj) collection
                         23.056  (± 8.7%) i/s -    114.000  in   5.006189s
Blueprinter collection(with Oj)
                         41.864  (±11.9%) i/s -    207.000  in   5.010554s

Comparison:
Light: collection through .to_json:       45.9 i/s
Blueprinter collection(with Oj):       41.9 i/s - same-ish: difference falls within error
ActiveModel::Serializers::JSON(with Oj) collection:       23.1 i/s - 1.99x  slower
AMS(without logging) (with Oj): collection:       18.4 i/s - 2.49x  slower
Surrealist: collection through Surrealist.surrealize_collection():       16.6 i/s - 2.77x  slower
Surrealist: collection through Surrealist::Serializer:       13.5 i/s - 3.39x  slower

Warming up --------------------------------------
AMS (associations): instance
                       512.000  i/100ms
Surrealist (associations): instance through .surrealize
                       710.000  i/100ms
Light (associations): instance through .to_json
                         2.379k i/100ms
Surrealist (associations): instance through Surrealist::Serializer
                       687.000  i/100ms
ActiveModel::Serializers::JSON (associations)
                         1.402k i/100ms
Blueprinter (associations)
                         2.035k i/100ms
Calculating -------------------------------------
AMS (associations): instance
                          5.878k (±17.0%) i/s -     28.672k in   5.074770s
Surrealist (associations): instance through .surrealize
                          7.019k (±18.5%) i/s -     33.370k in   5.028396s
Light (associations): instance through .to_json
                         28.010k (± 6.5%) i/s -    140.361k in   5.033427s
Surrealist (associations): instance through Surrealist::Serializer
                          7.819k (± 6.6%) i/s -     39.159k in   5.030462s
ActiveModel::Serializers::JSON (associations)
                         17.039k (± 7.6%) i/s -     85.522k in   5.049637s
Blueprinter (associations)
                         25.232k (± 6.5%) i/s -    126.170k in   5.022223s

Comparison:
Light (associations): instance through .to_json:    28009.7 i/s
Blueprinter (associations):    25231.8 i/s - same-ish: difference falls within error
ActiveModel::Serializers::JSON (associations):    17039.5 i/s - 1.64x  slower
Surrealist (associations): instance through Surrealist::Serializer:     7819.1 i/s - 3.58x  slower
Surrealist (associations): instance through .surrealize:     7018.8 i/s - 3.99x  slower
AMS (associations): instance:     5878.0 i/s - 4.77x  slower

Warming up --------------------------------------
AMS (associations): collection
                         1.000  i/100ms
Surrealist (associations): collection through Surrealist.surrealize_collection()
                         1.000  i/100ms
Light (associations): collection through .to_json
                         2.000  i/100ms
Surrealist (associations): collection through Surrealist::Serializer
                         1.000  i/100ms
ActiveModel::Serializers::JSON (associations): collection
                         1.000  i/100ms
Blueprinter (associations): collection
                         1.000  i/100ms
Calculating -------------------------------------
AMS (associations): collection
                          3.827  (± 0.0%) i/s -     19.000  in   5.016177s
Surrealist (associations): collection through Surrealist.surrealize_collection()
                          4.622  (±21.6%) i/s -     23.000  in   5.065690s
Light (associations): collection through .to_json
                         17.479  (±11.4%) i/s -     88.000  in   5.125929s
Surrealist (associations): collection through Surrealist::Serializer
                          4.481  (± 0.0%) i/s -     23.000  in   5.181694s
ActiveModel::Serializers::JSON (associations): collection
                         10.519  (±19.0%) i/s -     52.000  in   5.064718s
Blueprinter (associations): collection
                         15.987  (±18.8%) i/s -     76.000  in   5.048492s

Comparison:
Light (associations): collection through .to_json:       17.5 i/s
Blueprinter (associations): collection:       16.0 i/s - same-ish: difference falls within error
ActiveModel::Serializers::JSON (associations): collection:       10.5 i/s - 1.66x  slower
Surrealist (associations): collection through Surrealist.surrealize_collection():        4.6 i/s - 3.78x  slower
Surrealist (associations): collection through Surrealist::Serializer:        4.5 i/s - 3.90x  slower
AMS (associations): collection:        3.8 i/s - 4.57x  slower
```

### CODE REVIEW

File: lib/light_serializer/serializer.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'oj'
+require_relative 'hashed_object'
```
Comment (krim) [inline lib/light_serializer/serializer.rb]: может будем юзать `require`? Так будет потом легче не запутаться

File: lib/light_serializer/serializer.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+module LightSerializer
+  class Serializer
+    attr_reader :object
+
+    class << self
```
Comment (krim) [inline lib/light_serializer/serializer.rb]: давай сделаем отдельные методы `self.`?

File: lib/light_serializer/hashed_object.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      serializer.class.attributes.each_with_object({}) do |attribute, result|
+        obtain_values(attribute, object, result)
+      end
+    end
+
+    def obtain_values(attribute, object, result)
+      if attribute.is_a?(Hash)
+        values_from_nested_resource(attribute, object, result)
+      else
+        value = obtain_value(object, attribute)
```
Comment (krim) [inline lib/light_serializer/hashed_object.rb]: эту ветвь можно так же сделать отдельным методом - `value_from_current_resource`, как считаешь?

**Review by krim — APPROVED** [review]

File: lib/light_serializer/serializer.rb (hunk 1 above)
Comment (myxaluch) [reply lib/light_serializer/serializer.rb]: Да, согласен, поправлю)

File: lib/light_serializer/serializer.rb (hunk 2 above)
Comment (myxaluch) [reply lib/light_serializer/serializer.rb]: 👍

File: lib/light_serializer/hashed_object.rb (hunk 3 above)
Comment (myxaluch) [reply lib/light_serializer/hashed_object.rb]: да, отличный вариант, поправлю

**Pull request merged**